### DISCUSSION TITLE
[Feature] Transaction simulation doesn't work in (m)any tools

### DISCUSSION THREAD

**Discussion by aspiers** [celo-org/celo-blockchain#2291] [Feature requests (Needs)]
### Rationale

Hi all,

We noticed that when using any well-known developer tool which can simulate transactions, the simulation never works on Celo.  Examples include OpenZeppelin Defender and Tenderly.  This introduces considerable inconvenience and risk when preparing high value/impact transactions on mainnet.  Is there a reason for this limitation, and please can it be fixed?  We don't see this issue on any other chains.

### Implementation

Simulations should work just like they do on other standard EVM chains (Ethereum, Polygon etc.)

**Comment by aspiers** [comment]
Here's another example with Rainbow wallet:

![image](https://github.com/celo-org/celo-blockchain/assets/100738/4bd0f5f5-2f41-4bc7-8dea-aa89c19acb05)

**Comment by arthurgousset** [comment]
Not quite a solution for what you reported, but wanted to share that we have a PR in flight to improve transaction simulation in [`viem`](https://viem.sh/): 
- https://github.com/celo-org/viem/issues/11

cc @shazarre (for visibility)

**Comment by aspiers** [comment]
Here's an example in Defender 2.0:

![image](https://github.com/celo-org/celo-blockchain/assets/100738/b5d948ba-32f0-4921-b448-4ad9f5bcea49)

This is **severely** limiting our ability to safely interact with Celo.

**Comment by aspiers** [comment]
I found a workaround using [Foundry](https://getfoundry.sh/):

```
anvil --fork-url https://celo-mainnet.infura.io/v3/$API_KEY

# In another terminal
cast call -r https://celo-mainnet.infura.io/v3/$API_KEY --trace --data "$TX_DATA" --from $SENDER $TO_CONTRACT
```

Even better, piping the output through [rolod0x CLI](https://rolod0x.io/docs/user-manual.html#cli) can translate all the addresses in the trace output to human-readable labels, which makes the trace easy to understand.

That said, I still think it's super important for Celo to get simulation working in tools like OZ Defender.

**Reply by aspiers** [reply]
I found that sometimes this fails due to https://github.com/foundry-rs/foundry/issues/6855 and requires `foundryup -b master` to get the fix.

**Reply by arthurgousset** [reply]
Thanks for following up with the workaround and the detailed notes. This is super useful, really appreciate it!
